### DISCUSSION TITLE
Классы темплейтов

### DISCUSSION THREAD

**Discussion by furtivite** [doka-guide/platform#25] [Q&A] [answered]
Сейчас у нас есть базовый шаблон, в который встраиваются блоки с контентом страниц, выглядит как

```html
<html class="base">
    <head>
    </head>
    <body class="base__body">
        <!-- вставляется контент -->
        <div class="index">
            <!-- вставляется контент -->
        </div>
    </body>
</html>
```

Это увеличивает количество обёрток у блоков, и отходит от БЭМ, который предлагает использовать для таких случаев [миксы](https://ru.bem.info/methodology/key-concepts/#микс). 

Этот элемент по БЭМ стоило бы сверстать так:

```html
<html class="base">
    <head>
    </head>
    <body class="base__body index">
        <!-- вставляется контент -->
    </body>
</html>
```

Чтобы реализовать такой подход мы можем воспользоваться средствами nunjucks и создать переменную, которая будет передавать нужный нам класс, тогда запись будет выглядеть таким образом

базовый шаблон: 
```nunjucks
<html class="base">
    <head>
    </head>
    <body class="base__body {% if templateClass %}{{ templateClass }}{% endif %}">
        <!-- вставляется контент -->
    </body>
</html>
```

шаблон страницы:
```nunjucks
---
templateClass: index
---
<!-- вставляется контент -->
```

Мы используем небольшое количество шаблонов: 
- главная страница
- страница с aside
- навигационная страница
- сам aside

Считаю, что такой вариант записи был бы более подходящим, к тому же он чаще используется в остальных [основанных на 11ty проектах](https://www.11ty.dev/docs/starter/). 

Поскольку штука спорная, выносится на обсуждение.

**Comment by furtivite** [comment]
Начало обсуждения появилось [тут](https://github.com/Y-Doka/platform/pull/18#discussion_r597843348).

**Comment by pepelsbey** [comment, marked as answer]
> Это увеличивает количество обёрток у блоков, и отходит от БЭМ, который предлагает использовать для таких случаев миксы. Этот элемент по БЭМ стоило бы сверстать так:

БЭМ — это про блоки, элементы и модификаторы. Он же не БЭММ, правда? :) Миксы — это инструмент, который позволяет совместить на одном элементе свойства от разных блоков и элементов. Например, чтобы подмешать (смиксовать) к блоку какую-то геометрию от места, куда его вставляют.

Нет такого правила в БЭМе, что всё нужно миксовать чтобы только не было обёрток. Зато есть правило, точнее, практика, что явное лучше неявного. Смотрите:

Это базовый шаблон, который применяют все страницы:

```html
<!-- base.njk -->
<html class="base">
    <head>
    </head>
    <body class="base__body {% if templateClass %}{{ templateClass }}{% endif %}">
        <!-- вставляется контент -->
    </body>
</html>
```

Это шаблон конкретной страницы:

```html
<!-- index.njk -->
---
title: Главная
templateClass: index
tags: …
labels: …
---

<div class="index__header">
</div>
```

В чём проблема:

1. Имя блока `index` отсутствует во всех шаблонах, оно спрятано в мета-блоке. То есть `class="index"` в шаблонах не найти.
2. У блока `index` оторвали голову, у него нет понятного корня — его оторвали и положили в _другой шаблон_.

На мой взгляд, размазывать структуру блока по разным шаблонам — это плохая практика. Она делает код неявным.

## Так будет лучше

- Меньше условий
- Меньше ключей
- Блок сразу целиком

```html
<!-- base.njk -->
<html class="base">
    <head>
    </head>
    <body class="base__body">
        <!-- вставляется контент -->
    </body>
</html>
```

Это шаблон конкретной страницы:

```html
<!-- index.njk -->
---
title: Главная
tags: …
labels: …
---

<div class="index">
  <div class="index__header">
  </div>
</div>
```

**Reply by nlopin** [reply]
Стопроцентно за вариант Вадима. Это и в одном месте, и декларативно, и логики никакой вокруг нет

**Reply by Realetive** [reply]
Да, тут миксование ради меньшей вложенности — это «экономия на списках». Полезное из первого варианта можно вынести то, что если на какой-то странице может как-то отличаться дизайн элементов вне статьи (хэдер, футер, меню), то это можно будет сделать с помощью модификатора:

```html
<html class="base {% if templateClass %}base_route_{{ templateClass }}{% endif %}">
    <head>
    </head>
    <body class="base__body">
        <!-- вставляется контент -->
    </body>
</html>
```

```css
.base_route_about .base__header {
  margin-bottom: 30px;
}
```

**Reply by furtivite** [reply]
> Полезное из первого варианта можно вынести то, что если на какой-то странице может как-то отличаться дизайн элементов вне статьи (хэдер, футер, меню), то это можно будет сделать с помощью модификатора:
> 
> ```
> <html class="base {% if templateClass %}base_route_{{ templateClass }}{% endif %}">
>     <head>
>     </head>
>     <body class="base__body">
>         
>     </body>
> </html>
> ```
> 
> ```css
> .base_route_about .base__header {
>   margin-bottom: 30px;
> }
> ```

Ну вот такие модификаторы, действительно, полезны и будут применяться по необходимости.
